### PR TITLE
[Paladin] Basic T28 Support as well as T28 2p and 4p for Prot

### PR DESCRIPTION
**Pull request by Woliance**

### CODE REVIEW

File: engine/class_modules/paladin/sc_paladin.cpp [hunk 1]
```diff
@@ -3004,6 +3015,15 @@ void paladin_t::assess_damage( school_e school, result_amount_type dtype, action
...
 
+  // On a block event, trigger T28 4p if equipped
+  // todo: Woli -  Set bonus check
+  if ( ( s->block_result == BLOCK_RESULT_BLOCKED )  && ( sets->has_set_bonus( PALADIN_PROTECTION, T28, B4 ) ) &&
+       ( rng().roll( tier_sets.glorious_purpose_4pc->effectN( 1 ).percent() ) ) ) 
+  //)
+  {
+    trigger_t28_4p_pp( s );
+  }
+
```
Comment (Melekus) [inline engine/class_modules/paladin/sc_paladin.cpp]:
```suggestion
  // On a block event, trigger T28 4p if equipped
  // todo: Woli -  Set bonus check
  if ( ( s->block_result == BLOCK_RESULT_BLOCKED )  && sets->has_set_bonus( PALADIN_PROTECTION, T28, B4 ) 
        && rng().roll( tier_sets.glorious_purpose_4pc->effectN( 1 ).percent() ) ) 
  {
    trigger_t28_4p_pp( s );
  }

```
Trim some parenthesis, add a new line, remove some commented clutter

File: engine/class_modules/paladin/sc_paladin.cpp [hunk 2]
```diff
@@ -3490,7 +3510,12 @@ struct paladin_module_t : public module_t
 
     // 9.0 Paladin Night Fae
 
-    // Only create these if the player is a Night Fae Paladin or sets the option to get the buff.
+    // Only create these if the player is a Night Fae Paladin or sets the option to get the 
+    // 
+    // 
+    // 
+    // 
+    //
```
Comment (Melekus) [inline engine/class_modules/paladin/sc_paladin.cpp]:
```suggestion
    // Only create these if the player is a Night Fae Paladin or sets the option to get the buff.
```
undo weird changes to comments

File: engine/class_modules/paladin/sc_paladin_protection.cpp [hunk 3]
```diff
@@ -751,6 +771,20 @@ void paladin_t::trigger_holy_shield( action_state_t* s )
   active.holy_shield_damage -> schedule_execute();
 }
 
+void paladin_t::trigger_t28_4p_pp( action_state_t* s )
+{
+  // escape if we don't have t28 4p
```
Comment (scamille) [inline engine/class_modules/paladin/sc_paladin_protection.cpp:776]: Please remove the comment lines here. Alternatively, you could move all of the conditions you have at the caller into this function.

File: engine/class_modules/paladin/sc_paladin_protection.cpp [hunk 4]
```diff
@@ -751,6 +771,20 @@ void paladin_t::trigger_holy_shield( action_state_t* s )
...
+  // if ( !talents.holy_shield->ok() )
+  //  return;
+
+  // sanity check - no friendly-fire
+  if ( !s->action->player->is_enemy() )
```
Comment (scamille) [inline engine/class_modules/paladin/sc_paladin_protection.cpp:781]: I think you should use s->target->is_enemy() here.

File: engine/class_modules/paladin/sc_paladin_protection.cpp [hunk 5]
```diff
@@ -751,6 +771,20 @@ void paladin_t::trigger_holy_shield( action_state_t* s )
...
+    return;
+
```
Comment (scamille) [inline engine/class_modules/paladin/sc_paladin_protection.cpp:783]: You can add the following here, makes debugging easier if your trigger and proc action creation conditions ever get out of sync:
`assert( active.t28_4p_pp != nullptr );`

File: engine/class_modules/paladin/sc_paladin_protection.cpp [hunk 6]
```diff
@@ -784,6 +818,11 @@ void paladin_t::create_prot_actions()
     active.judgment = new judgment_prot_t( this );
 
   active.divine_resonance = new avengers_shield_dr_t( this );
+
+  if ( specialization() == PALADIN_PROTECTION )
```
Comment (scamille) [inline engine/class_modules/paladin/sc_paladin_protection.cpp:822]: I would replace the spec check with
`if (sets->has_set_bonus( PALADIN_PROTECTION, T28, B4 ) )`

**Review by scamille — COMMENTED** [review]
Some suggestions, even though this is already merged.

**Pull request merged**